Add eMagLS for EMA and extensive project updates

HaHeho opened the pull request:
TODOs before merge:
- [x] Perform merge of Tommi's changes into this branch
- [x] Cleanup leftovers functions from preliminary EMA implementations
- [x] Add ambisonic encoding toolbox as git-submodule (for `getEMagLsFiltersEMAinSH()`)
- [x] Add headers with documentation and literature references to all new functions
- [x] Update headers in all existing functions
- [x] Add reference to EMA publication to README
- [x] Move Changelog from README to separate file
- [x] Discuss keeping the `getMagLsArrayDiffuseFilter.m` and `getMagLsSphericalHeadFilter.m` functions
- [x] Discuss keeping the complex implementations in all filter functions
- [x] Discuss keeping the optional `shDefinition` and `shFunction` parameters in all filter functions
- [x] Update the example file not to specify the `shDefinition` and `shFunction` parameters (i.e. use the default values)
- [x] Update the filter functions to use the default value "false" for the `applyDiffusenessConst` parameter
- [x] Update the example file not to specify the `applyDiffusenessConst` parameter (i.e. use the default value)
- [x] @thomasdeppisch validate the updated rendering filters against the currently published implementation (the filters should be very similar with minor differences due to numerical precision)
- [x] Update example to remove download and re-add audio source file
- [x] Update function headers to the year 2023

TODOs after merge:
- [x] Rename `master` into `main` branch
- [x] Update all function headers and documentation referencing the LICENSE file with the updated `main` URL
- [x] Add git tags to some milestone stages of the repository and its updates
- [x] Add GitHub releases according to the chosen tags

HaHeho committed:
Fix function endings

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Remove default HRIR dataset
(update example script to download data on demand)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Remove default SMA recording dataset
(update example script to download data on demand)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update .gitignore

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update example script to be more verbose
(reintroduce length limitation of the rendered SMA recording)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update example script with variable to configure the application of the MagLS diffuseness constraint

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update example script to export reference results for LS, MagLS, eMagLS and eMagLS2 filters
(also add the respective reference results)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update example script to add global configuration variables

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update example script to add verification of rendering filters against provided reference

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to provide SH basis type (real or complex) as option parameter

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to provide utilized constants at the top

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to provide SH basis implementation function as an optional parameter
(see `testEMagLs.m` for a usage example)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update README.md with changelog

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix functions to use case-insensitive string comparisons for parameters

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to export audio rendering results as an option

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to add used SH basis type in exported file names
(update file names of respective reference results)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFilters.m` and `getEMagLs2Filters.m` to slightly improve computation time

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to use `ifft()` without the forced symmetric parameter
(fix to yield real or complex filters depending on the basis type)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to accept tolerance when verifying rendering filters against the provided reference

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getMagLsFilters.m` to streamline conjugate and transpose operations
(results in neglectable maximum normalized absolute difference of 1e-15)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getLsFilters.m` to streamline conjugate and transpose operations
(results in neglectable maximum normalized absolute difference of 1e-16)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFilters.m` to streamline conjugate and transpose operations
(results in neglectable maximum normalized absolute difference of 1e-15)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLs2Filters.m` to streamline conjugate and transpose operations
(results in neglectable maximum normalized absolute difference of 1e-14)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getSMAIRMatrix.m` to streamline conjugate and transpose operations

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update .gitmodules to add Array-Response-Simulator again

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getRadialFilter.m` and `getSMAIRMatrix.m` to use `sphModalCoeffs()` from Array-Response-Simulator
(therefore also remove own implementations of spherical hankel and bessel functions)
(this causes the resulting eMagLS and eMagLS2 rendering filters to be slightly different at 0 Hz and the Nyquist bin)
(therefore the reference for eMagLS and eMagLS2 rendering filters are updated for verification)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to use a different SH basis implementation more easily

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `binauralDecode.m` to warn when rendering discards imaginary signal parts (may occur for complex SH basis functions)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update scripts with minor fixes

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to export binaural renderings as FLAC with 24bit resolution

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to verify the SH convention "wikipedia" against the "real" reference

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to compute 0 Hz bin not separately
(this causes the resulting eMagLS and eMagLS2 rendering filters to be different)
(therefore the reference for eMagLS and eMagLS2 rendering filters are updated for verification)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to alternatively evaluate spectral difference in verification of rendering filters against provided reference

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to work for complex SH conventions by computing double-sided spectra
(required in `getMagLsFilters.m` and `getEMagLsFilters.m` but not in `getEMagLs2Filters.m`)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to export binaural renderings as WAV with 64bit resolution

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getMagLsFilters.m`, `getEMagLsFilters.m` and `getEMagLs2Filters.m` to use a transition frequency of 500 Hz times SH order instead of always 2000 Hz
(changes the signature of `getEMagLs2Filters.m` to require the desired SH order)
(yields no changes for the example Eigenmike configuration)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to fix generation of path name strings on Windows

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `applyRadialFilter.m` to clarify application of casual version of radial filters

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getSMAIRMatrix.m` to determine the simulation order based on the configuration's aliasing frequency
(this causes the resulting eMagLS and eMagLS2 rendering filters to be different in a non-meaningful way at very high frequencies)
(therefore the reference for eMagLS and eMagLS2 rendering filters are updated for verification)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getMagLsFilters.m`, `getEMagLsFilters.m` and `getEMagLs2Filters.m` to use delays with subsample precision and restore original inter-aural group delay difference after magnitude least-squares optimization
(this causes the resulting MagLS, eMagLS and eMagLS2 rendering filters to be different in a non-meaningful way at very high frequencies)
(therefore the reference for MagLS, eMagLS and eMagLS2 rendering filters are updated for verification)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add reference rendering filters for "complex" SH convention for verification

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add `getEMagLsFiltersEMA.m` to generate eMagLS rendering filters for equatorial microphone arrays (EMA)
(add the respective `getCH.m` and `ch_stackOrder.m` functions)
(this is does not have an example configuration in `testEMagLs.m` yet)
(the rendering filters for "complex" SH basis types do not function as intended yet)
(the implementation of the diffuseness constraint has not been verified yet)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFiltersEMA.m` to remove unused array grid zenith coordinates
(add non-functioning example call to `testEMagLs.m`)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to extract generation of time domain window for resulting rendering filters
(add the respective `getFadeWindow.m` function)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to simplify syntax of singular value decomposition

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFiltersEMA.m` to temporarily use only horizontal HRIRs for magnitude optimization
(spherical HRIR data will be SH sub-sampled to a horizontal grid)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFiltersEMA.m` SVD regularization parameter to yield more sensible rendering filters that
(better behaviour in time domain at the cost of small deviations in magnitude response)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix `getEMagLsFiltersEMA.m` parsing of default CH basis function handle

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Rename `getEMagLsFiltersEMA.m` into ``getEMagLsFiltersEMAinEHtoSH.m` to reflect its function

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add `getEMagLsFiltersEMAinCH.m` and `getEMagLsFiltersEMAinCHtoSH.m` with different preliminary implementations of EMA rendering filters

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update README.md

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix functions to calculate the filters at the specified oversampled length</br>
(now the oversampling actually uses double the specified length up to a maximum of 2048 samples)
(this causes all the resulting rendering filters to be different at high frequencies in a non-meaningful way)
(therefore the reference for all the rendering filters are updated for verification)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix `getEMagLsFiltersEMAinEHtoSH.m` to reflect switch from "soundfieldsynthesis" scripts to updated "Ambisonic Encoding" toolbox

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFiltersEMAinCHtoSH.m` and `getEMagLsFiltersEMAinEHtoSH.m` to extract helper functions</br>
(add the respective `ch_fromShIds.m` and `eh_fromShIds.m` functions)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFiltersEMAinCH.m`, `getEMagLsFiltersEMAinCHtoSH.m` and `getEMagLsFiltersEMAinEHtoSH.m` to improve computation speed
(extract feasible matrix multiplications to be pre-computed before the loop)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update function documentation

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add functions `getMagLsSphericalHeadFilter.m` and `getMagLsArrayDiffuseFilter.m` to provide direction-independent array equalization filters that can be useful in combination with MagLS

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to calculate the HRIR group delay in a simpler way without meaningful changes in the result
(this causes the resulting MagLS, eMagLS and eMagLS2 rendering filters to be different in a non-meaningful way at very high frequencies)
(therefore the reference for MagLS, eMagLS and eMagLS2 rendering filters are updated for verification)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update preliminary EMA functions to subsample a spherical HRIR to a circular grid at lower resolution
(this should improve the resulting rendering filters for lower-resolution input HRIRs)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Revert bf61bdf97d9444234efadec140300c474a8428d4
(this turned out to be a bad idea at medium to high SH orders)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add `getEMagLsFiltersEMAinSH.m` to provide the implementation of EMA rendering filters in spherical harmonics

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getMagLsSphericalHeadFilter.m` and `getMagLsArrayDiffuseFilter.m` to remove amplitude normalization and yield no magnitude changes at low frequencies

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFiltersEMAinSH.m` to use existing EMA function from the Ambisonic Encoding toolbox

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to improve computation speed by pre-allocating complex matrices when needed

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getSMAIRMatrix.m` to work for higher requested array orders than estimated simulation order

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix `getEMagLsFiltersEMAinSH.m` vertical rotation direction (it was inverted before)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFiltersEMAinSH.m` to improve computation speed by performing all sound field rotations in one operation

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update functions to use a transition frequency of at least 1 kHz (this makes a difference only for SH order 1)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix `getEMagLs2Filters.m` when applying the diffuseness constraint

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to yield reference names including the status of the diffuseness constraint
(rename the respective reference rendering filter files)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add reference rendering filters with diffuseness constraint for verification

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update SMA functions to improve syntax for diffuseness constraint (the resulting filters are identical)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix EMA functions to implement the diffuseness constraint

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update eMagLS functions with comment that diffuseness constraint causes slight magnitude deviations at low frequencies

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Remove outdated functions with preliminary implementations of EMA rendering filters

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getMagLsFilters.m` to improve computation speed by matricizing all operations for both ears
(something similar should also be implemented for the other functions in the future)
(this causes the resulting rendering filters with diffuseness constraint to be different at high frequencies in a non-meaningful way)
(therefore the reference for the rendering filters with diffuseness constraint are updated for verification)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Merge remote-tracking branch 'origin/master' into reference_update

# Conflicts:
#	.gitignore
#	.gitmodules
#	dependencies/applyRadialFilter.m
#	dependencies/getRadialFilter.m
#	dependencies/getSMAIRMatrix.m
#	getEMagLs2Filters.m
#	getEMagLsFilters.m
#	getLsFilters.m
#	getMagLsFilters.m
#	testEMagLs.m

HaHeho committed:
Cleanup after merge

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `applyRadialFilter.m` to perform causalization and fading analogous to other filtering functions

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Remove outdated functions that were used only in preliminary implementations of EMA rendering filters

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update header documentation of all functions

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add `ambisonic-encoding` as git submodule (required for `get_sound_field_sh_coeffs_from_ema_t()` in `getEMagLsFiltersEMAinSH.m`)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `README.md` with references to the EMA publication

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add `CHANGELOG.md` with changelog moved from `README.md`

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update header documentation in filter functions

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

thomasdeppisch commented on dependencies/Spherical-Harmonic-Transform:
This does not seem right

thomasdeppisch commented on testEMagLs.m:
```diff
@@ -9,40 +9,222 @@
+[smaRecordingLength, smaRecordingFile, smaRecordingUrl] = deal( ...
+    13.5, ... % in seconds
+    'resources/Acappella_Eigenmike_Raw_32ch.wav', ...
+    'https://zenodo.org/record/3477602/files/09%203D-MARCo%20Samples_Acappella.zip');
```
This unfortunately downloads a lot of data and takes forever, maybe rather go back to old version with small file directly available in repository?

HaHeho replied:
For now, I added a comment to the command line that the download might take a while on the first run.

HaHeho committed:
Fix missing `Spherical-Harmonic-Transform` git submodule

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

thomasdeppisch commented on testEMagLs.m:
```diff
@@ -9,40 +9,222 @@
+            fprintf('done.\n');
+        
+            refFile = sprintf(refFiles, refShDefinition, refMagLS);
+            fprintf('Verifying LS rendering filters against "%s" ... ', refFile);
```
Copy/paste error in fprintf (should be magLS instead of LS)

thomasdeppisch commented on testEMagLs.m:
```diff
@@ -9,40 +9,222 @@
+            fprintf('done.\n');
+        
+            refFile = sprintf(refFiles, refShDefinition, refeMagLS);
+            fprintf('Verifying LS rendering filters against "%s" ... ', refFile);
```
Copy/paste error in fprintf

thomasdeppisch commented on testEMagLs.m:
```diff
@@ -9,40 +9,222 @@
+            fprintf('done.\n');
+        
+            refFile = sprintf(refFiles, refShDefinition, refeMagLS2);
+            fprintf('Verifying LS rendering filters against "%s" ... ', refFile);
```
Copy/paste error in fprintf

HaHeho committed:
Update `testEMagLs.m` documentation

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

thomasdeppisch commented on testEMagLs.m:
```diff
@@ -9,40 +9,222 @@
+DO_PLAYBACK_RENDERING = true;
 
-% load HRIR set (see https://zenodo.org/record/3928297)
-load HRIR_L2702.mat
```
This requires a miro.m header file, we should at least make aware of with a comment

HaHeho committed:
Update `testEMagLs.m` to download the MIRO class if it is not available for loading the default HRIR data set

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getSMAIRMatrix.m` to export data for debugging of system-dependent differences in resulting rendering filters

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho commented on dependencies/getSMAIRMatrix.m:
```diff
@@ -79,68 +82,85 @@
-    
\ No newline at end of file
+
+    % TODO: resolve system-dependent differences in resulting eMagLS and eMagLS2 filters
```
This can be used to export the data on different systems and compare it directly. My data is also contained in 6a2fc2bdd6a823ff977f4fefa12279c32a900aab.

thomasdeppisch commented on dependencies/binauralDecode.m:
```diff
@@ -18,9 +21,10 @@
-        in = rotateHOA_N3D(in, horRotAngleRad*180/pi, 0, 0);
+        % TODO: This function requires 
+        %       https://github.com/polarch/Higher-Order-Ambisonics.git
+        in = rotateHOA_N3D(in, rad2deg(horRotAngleRad), 0, 0);
```
The function requires the Higher Order Ambisonics toolbox but it is not called in the current test script, so I would say we keep it as is.

HaHeho committed:
Update `getEMagLsFilters.m` to export data for debugging of system-dependent differences in resulting rendering filters

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

thomasdeppisch commented on dependencies/getCH.m:
```diff
@@ -0,0 +1,19 @@
+% Hannes Helmholz, 2022
+
+if strcmpi(shDefinition, 'complex')
+    m = ch_stackOrder(order);
```
Where is this function defined?

HaHeho committed:
Update `getEMagLsFilters.m` with temporary fix leaving out the DC bin

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getEMagLsFilters.m` to export data for debugging of system-dependent differences in resulting rendering filters

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Revert "Update `getSMAIRMatrix.m` to export data for debugging of system-dependent differences in resulting rendering filters"

This reverts commit 6a2fc2bd

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

thomasdeppisch commented on dependencies/getRadialFilter.m:
```diff
@@ -1,61 +1,69 @@
-        error('unkown radialFilter parameter');
-        
+    if ~mod(nfft, 2) % is even
+        radFilts(end, :) = real(radFilts(end, :)); % Nyquist bin
```
I think here it would be better to take abs(radFilts(end, :))

thomasdeppisch commented on getEMagLs2Filters.m:
```diff
@@ -15,117 +20,95 @@
-W_LS_r = zeros(nfft, numMics);
-W_MLS_l = zeros(nfft, numMics);
-W_MLS_r = zeros(nfft, numMics);
-for k = 1:numPosFreqs % leave out first bin, here bn = 0
```
apply same changes as in eMagLs (k=2:end and DC mod)

thomasdeppisch commented on getEMagLsFilters.m:
```diff
@@ -17,178 +22,190 @@
 end
 
+% TODO: resolve system-dependent differences in resulting eMagLS and eMagLS2 filters
+var_name = 'Y_reg_inv_exp';
```
remove this

thomasdeppisch commented on getEMagLsFiltersEMAinCH.m:
```diff
@@ -0,0 +1,220 @@
+
+W_MLS_l = zeros(nfft, numHarmonics, 'like', HL);
+W_MLS_r = zeros(nfft, numHarmonics, 'like', HL);
+for k = 1:numPosFreqs
```
apply same changes as in eMagLs (k=2:end and DC mod)

thomasdeppisch commented on getEMagLsFiltersEMAinSH.m:
```diff
@@ -0,0 +1,242 @@
+
+W_MLS_l = zeros(nfft, numHarmonics, 'like', HL);
+W_MLS_r = zeros(nfft, numHarmonics, 'like', HL);
+for k = 1:numPosFreqs
```
apply same changes as in eMagLs (k=2:end and DC mod)

thomasdeppisch commented on testEMagLs.m:
```diff
@@ -9,40 +9,232 @@
+[hrirFile, hrirUrl] = deal('resources/HRIR_L2702.mat', ...
+    'https://zenodo.org/record/3928297/files/HRIR_L2702.mat');
+
+[smaRecordingLength, smaRecordingFile, smaRecordingUrl] = deal( ...
```
remove

HaHeho committed:
Revert "Update `getEMagLsFilters.m` to export data for debugging of system-dependent differences in resulting rendering filters"

This reverts commit 4b8a3f46

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

thomasdeppisch commented on testEMagLs.m:
```diff
@@ -9,40 +9,232 @@
-%% sh transform and radial filter (for LS and conventional MagLS)
-E = getSH(shOrder, [micGridAziRad, micGridZenRad], 'real');
-shRecording = smaRecording * pinv(E).';
+if DO_VERIFY_REFERENCE
```
remove verification?

thomasdeppisch commented on testEMagLs.m:
```diff
@@ -9,40 +9,232 @@
+end
+
+%% replace reference filters
+if DO_OVERRIDE_REFERENCE
```
remove reference

HaHeho committed:
Update eMagLS functions to manually set the DC bin after calculation
(therefore the reference for the rendering filters are updated for verification)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update eMagLS functions documentation when manually setting the DC bin after calculation

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to not perform validation of the rendering filters against the reference by default

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update eMagLS functions with error message when applying the diffuseness constraint with "complex" SH conventions (this is not implemented yet)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `getRadialFilter.m` to set the Nyquist bin for even-length filters after calculation

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix `getCH.m` to add missing function `ch_stackOrder()` (was accidentally removed before)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Fix `getEMagLs2Filters.m` to parse the number of optional function parameters correctly

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to remove the `shFunction` and `shDefinition` parameters
(therefore the default values "@getSH" and "real" of the rendering filter functions will be used)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to remove the `shFunction` and `shDefinition` parameters
(therefore the default values "@getSH" and "real" of the rendering filter functions will be used)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update `testEMagLs.m` to remove the `applyDiffusenessConst` parameter
(therefore the default value "false" of the rendering filter functions will be used)

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Add `resources/Acappella_Eigenmike_Raw_32ch_short.wav` again to not download a large ZIP file when executing the example for the first time

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
Update function header documentation and year

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho committed:
fixup! Add `resources/Acappella_Eigenmike_Raw_32ch_short.wav` again to not download a large ZIP file when executing the example for the first time

Signed-off-by: HaHeho <hannes.helmholz@chalmers.se>

HaHeho commented:
PR is merged. :) @thomasdeppisch, I think only you can rename the default branch of the repository to `main`. This would be good to do directly in the repository settings according to https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch.

Afterward, I will add a few release tags and perform the remaining steps.

thomasdeppisch commented:
Great, many thanks for your efforts! :) I'll look into the renaming now

thomasdeppisch commented:
I just renamed the master to main and cloned the repo and everything works as expected :)
We could think about adding an EMA rendering example to show how that works.

HaHeho commented:
> I just renamed the master to main and cloned the repo and everything works as expected :) We could think about adding an EMA rendering example to show how that works.

Nice. Yes and EMA rendering example would be nice to have as well. We will need a source audio recording for that. Ideally, we could even have the same scenario captured with an SMA and an EMA. Maybe something from the latest TU Berlin data set?

thomasdeppisch commented:
> > I just renamed the master to main and cloned the repo and everything works as expected :) We could think about adding an EMA rendering example to show how that works.
> 
> Nice. Yes and EMA rendering example would be nice to have as well. We will need a source audio recording for that. Ideally, we could even have the same scenario captured with an SMA and an EMA. Maybe something from the latest TU Berlin data set?

That’s a great idea!
